lenses.md

```diff
@@ -0,0 +1,222 @@
+
+## Temporary stop-gap measures
+
+If a codebase is ubiquitously using lenses as getters, then rewriting it to use the builtin syntax instead will take time. It might make sense to replace some _very_ commonly used lenses with monomorphic implementations. Although, in a codebase defining hundreds of record types, each of them with a few fields, there is most likely no small group of lenses whose monomorphization makes a difference.
+
+Fortunately, there's one significant win we can achive without changing too much code at all. The current implementation of `lens` is pretty far away from the implementation of `_field1`. If we move `lens` closer to `_field1`, we arrive at
+```haskell
+fastLens: forall x r a. HasField x r a => Lens r a
+fastLens f r = fmap (\x -> setField @x x r) (f (getField @x r))
+```
+Benchmarking this implementation gives us
+
+|`x`|`T(x)`|`(T(x) - T(benchNoop)) / 100,000`|
+|-|-|-|
+|`benchFastLens`|128.9 ms|1178 ns|
+
+These numbers are still not great, but they are at least a 1.5x speedup compared to the implementation of `lens`.
+
+## Chains of record accesses
+
+So far, our benchmarks were only concerned with accessing one field in one record. A pattern that occurs quite frequently in practice are nested records and chains of record accesses, as in
+```
+r.field1.field2.field3
+```
+With the builtin syntax, every record access you attach to the chain is as (in)expensive as the first record access. Benchmarks confirm this linear progression. However, we could _easily_ make every record access after the first one in a chain _significantly_ faster in the DAML interpreter.
+
+There's a similar linear progression when using `get` and `fastLens`. Unfortunately, we have _no chance_ of optimizing chains of record accesses in any way since they are completely intransparent to the compiler and the interpreter.
+
+## Conclusion
+
+I think the numbers say everything there's to say:
+
+|method|time (in ns)|slowdown vs builtin|
+|-|-|-|
+|builtin|46|1x|
+|monomorphic lens|814|17.7x|
+|polymorphic lens|1178|25.6x|
+|`lens`|1776|38.6x|
+
+In view of these numbers, I would recommend to everybody who cares about performance to use DAML's builtin syntax for accessing record fields!
+
+Only focussing on getters might be modestly controversial since lenses also serve a purpose as setters. I expect the differences in performance between DAML's builtin syntax for updating record fields
+```
+r with field1 = newValue
+```
+and using `set` and a lens to update a single field in a single record to be in the same ballpark as for getters. When using the builtin syntax to updat multiple fields in the same record, the DAML interpreter already performs some optimizations to avoid allocating intermediate records. Such optimizations are impossible with lenses.
```

Comment:
```suggestion
and using `set` and a lens to update a single field in a single record to be in the same ballpark as for getters. When using the builtin syntax to update multiple fields in the same record, the DAML interpreter already performs some optimizations to avoid allocating intermediate records. Such optimizations are impossible with lenses.
```